### PR TITLE
8312125: Refactor CDS enum class handling

### PR DESCRIPTION
**Pull request by iklam**
Please review this simple refactoring of the CDS code for handling enum classes. The code is moved to new files cdsEnumKlass.cpp/hpp. There's otherwise no change.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8312125](https://bugs.openjdk.org/browse/JDK-8312125): Refactor CDS enum class handling (**Sub-task** - P4)

### Reviewers
 * [Matias Saavedra Silva](https://openjdk.org/census#matsaave) (@matias9927 - **Reviewer**)
 * [Calvin Cheung](https://openjdk.org/census#ccheung) (@calvinccheung - **Reviewer**) 🔄 Re-review required (review applies to [49dc109e](https://git.openjdk.org/jdk/pull/20013/files/49dc109e2423a31496a2ec14c3cf4ab1e40e3361))

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20013/head:pull/20013` \
`$ git checkout pull/20013`

Update a local copy of the PR: \
`$ git checkout pull/20013` \
`$ git pull https://git.openjdk.org/jdk.git pull/20013/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20013`

View PR using the GUI difftool: \
`$ git pr show -t 20013`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20013.diff">https://git.openjdk.org/jdk/pull/20013.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20013#issuecomment-2206808683)

### CODE REVIEW

File: src/hotspot/share/cds/cdsEnumKlass.cpp [hunk 1]
```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright (c) 2023, Oracle and/or its affiliates. All rights reserved.
```
Comment (matias9927) [inline src/hotspot/share/cds/cdsEnumKlass.cpp:2]: Please update the copyrights on the new files.

File: src/hotspot/share/cds/cdsEnumKlass.cpp [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+// ModuleLayer::EMPTY_LAYER, for example).
+//
+// Instead, we archive all static field of such Enum classes. At runtime,
+// HeapShared::initialize_enum_klass() skips the <clinit> method and instead pulls
+// the static fields out of the archived heap.
+void CDSEnumKlass::handle_enum_obj(int level,
+                                   KlassSubGraphInfo* subgraph_info,
+                                   oop orig_obj) {
+  assert(level > 1, "must never be called at the first (outermost) level");
+  assert(is_enum_obj(orig_obj), "must be");
```
Comment (matias9927) [inline src/hotspot/share/cds/cdsEnumKlass.cpp:68]: Is this assert redundant? You check for this before you make the call to `handle_enum_obj()` below. I think you can move the if statement in here.

**Review by matias9927 — CHANGES_REQUESTED** [review]
I have two comments but overall this looks good!

File: src/hotspot/share/cds/cdsEnumKlass.cpp (hunk 2 above)
Comment (iklam) [reply src/hotspot/share/cds/cdsEnumKlass.cpp:68]: The assert declares the fact that the caller must check that orig_obj is an enum before calling this function. This guards against inadvertent changes that might accidentally drop the "if" check in the caller.

File: src/hotspot/share/cds/cdsEnumKlass.cpp (hunk 1 above)
Comment (iklam) [reply src/hotspot/share/cds/cdsEnumKlass.cpp]: I changed the dates to `2023, 2024` as these files were initially published in the leyden repo last year.

**Review by matias9927 — APPROVED** [review]
Thanks for the changes and clarification!

File: src/hotspot/share/cds/cdsEnumKlass.cpp [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+    if (fs.access_flags().is_static()) {
+      int root_index = info->enum_klass_static_field_root_index_at(i++);
+      fieldDescriptor& fd = fs.field_descriptor();
+      assert(fd.field_type() == T_OBJECT || fd.field_type() == T_ARRAY, "must be");
+      mirror->obj_field_put(fd.offset(), HeapShared::get_root(root_index, /*clear=*/true));
+    }
+  }
+  return true;
+}
+#endif
```
Comment (calvinccheung) [inline src/hotspot/share/cds/cdsEnumKlass.cpp:136]: Suggestion: `#endif INCLUDE_CDS_JAVA_HEAP`

**Review by calvinccheung — APPROVED** [review]
Refactoring looks good. I have one suggestion.

**Comment by iklam** [comment]
Thanks @matias9927 @calvinccheung for the review
/integrate

**Review by matias9927 — APPROVED** [review]
Updates look good

**Pull request closed without merging**